[sheenkerr] iP

sheenkerr opened the pull request:
# Future You

Future You is a task list chat bot that frees you from having to remember things you need to do. It's,

1. text-based
2. easy to learn
3. let it manage your tasks for you 👍

And it is [**FREE**](https://github.com/sheenkerr/ip)!

## Features:

- [x] Managing tasks  
- [x] Managing deadlines
- [ ] Sorting tasks (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Commit:
Changed name from Duke to Future You everywhere

Changed name from Duke to Future You in all files & added ASCII art acknowledgement in docs/README.md

Commit:
Added Rename, Greet, Exit

Commit:
Added Echo Feature

Commit:
Changed Echo Feature into add to list feature

Commit:
Added Task Class

Commit:
Added mark task functionality

Commit:
Added Deadline and Events Classes

Commit:
Added getType() to all classes & print function to Task class

Commit:
Changed add to list using todo keyword

Refactored code using new function addTask and print task details using function print(). Started on error handling, when user does not key in valid command, it will just echo the user input."

Commit:
Added deadline adding feature

Added new addDeadline function. Overriden print() in Deadline to include the (by: <deadline>) inside.

Commit:
Added event adding feature

Added new addEvent function. Overriden print() in Events to include the (from: <startDateTime> to: <endDateTime>) inside.

Commit:
Added Error Handling for invalid inputs

Commit:
Added support for deleting tasks

Commit:
Started on Automated Text UI Testing

Commit:
Add UI class and extract UI functions

Commit:
Add Parser class with parse command functions

Commit:
Add new FutureYouException class

Commit:
Add TaskList class with functions extracted

Commit:
Update Parser class to work with CRUD functions of TaskList class

Commit:
Update UI class to use Parser class functions

Commit:
Refactor FutureYou and fixed new line bug not showing

Commit:
Create new folders to start conversion to packages

Commit:
Format the import and package statements

Commit:
Add new toString function for storage formating

Commit:
Change "\n" to System.lineSeparator() so that it works on multiple OS

Commit:
Add Storage class with full functionality

Commit:
Update TaskList class to work with Storage

Commit:
Fix bugs in TaskList class when reading from file

Commit:
Update FutureYou to work with new Storage class

Commit:
Update Ui class to work with new FutureYou class

Commit:
Update all classes to use LocalDateTime for Deadline Tasks

Commit:
Fix bug causing incorrect LocalDate format from saving

Commit:
Fix bug causing addiitonal space saved in txt file

Commit:
Update Deadline class to use LocalDateTime

Commit:
Update Events class to use LocalDateTime

Commit:
Merge branch 'add-gradle-support'

Commit:
Add Gradle to run program

Commit:
Add two JUnit test cases for creation of new Task

Commit:
Update Gradle to package app as JAR file

Commit:
Update retrieval function for task name

Commit:
Add find function to list all tasks that contains user input

Commit:
Add find functionality to Ui and Parser

Commit:
Add javadoc for Deadline class

Commit:
Add javadoc for Events class

Commit:
Add javadoc for Task class

Commit:
Reformat switch case to match coding standard

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Merge branch 'branch-Level-9'

PeterHW963 commented on src/main/java/futureYou/Storage.java:
```diff
@@ -0,0 +1,133 @@
+     * Ensures the file at the specified file path exists.
+     * If the file does not exist, it creates the necessary parent directories and
+     * the file.
+     * If the file already exists, no action is taken.
```
I think the JavaDocs here is incomplete? (i.e. no @param line)

PeterHW963 commented on src/main/java/futureYou/Storage.java:
```diff
@@ -0,0 +1,133 @@
+    private Task parseTask(String taskString) {
+        String[] stringParts = taskString.trim().split("\\|");
+
+        String typeOfTask = stringParts[0].trim();
```
I really like how you named each `stringParts` array elements, I should start doing this in my code as well

PeterHW963 commented on src/main/java/futureYou/task/Task.java:
Perhaps a more intuitive name for this file would be `Todo`? If I saw your repository file names without looking at the code, I would assume that `Task` is an abstract parent class. 

Alternatively, maybe you can rename the other task classes to be `DeadlineTask` and `EventTask`

PeterHW963 replied:
PS I just realized that your Task class is also parent class for both deadline and event, so my suggestions are either:
1.  Make `Task` class abstract, and have `TodoTask`, `DeadlineTask`, `EventTask` as child classes of the `Task` class, or
2. Rename `Deadline` and `Event` to `DeadlineTask` and `EventTask` and explain in the JavaDoc for the `Task` class that it also contains `TodoTask` behaviours.

PeterHW963 commented on src/main/java/futureYou/task/Task.java:
```diff
@@ -0,0 +1,87 @@
+     * @return A formatted string for displaying the task.
+     */
+    public String print() {
+        String cross = (this.getTaskStatus() ? "X" : " ");
```
Again, I like how you separate these things for better clarity 😄

PeterHW963 commented on src/test/java/futureYou/DeadlineTest.java:
```diff
@@ -0,0 +1,27 @@
+        LocalDateTime deadline = LocalDateTime.of(2024, 9, 05, 14, 0);
+        Deadline newDeasdlineTask = new Deadline(taskName, deadline);
+
+        assertEquals(deadline.format(format), newDeasdlineTask.getDeadline());
```
typo here? Does this test run?

PeterHW963 reviewed (commented):
Solid overall :) 
Am learning a lot about how to structure my own program from seeing your code

Commit:
Add checkstyle files and in gradle

Commit:
Update package name

Commit:
Update all files to fit with checkStyle

Commit:
Update Parser & Storage class to fit with checkStyle

Commit:
Add GUI iteration one

Commit:
Add GUI iteration two

Commit:
Add GUI iteration three

Commit:
Change code to use .fxml files

Commit:
Change code to futureYou instead of Duke

Commit:
Update code to work with GUI

Solomon0519 commented:
Code is clear, variables make it clear what their purpose is, learnt a lot from your code!

Commit:
Update initializeFile() to use guard if
Refactored the initializeFile() method to use a guard clause for better readability and to reduce the nesting of conditional statements.

This change helps make the code easier to follow by returning early when conditions are not met, adhering to clean code principles.

Commit:
Update formating for files

Commit:
Extract common task addition message to newTaskAddedMsg() method
Refactored the taskList logic by extracting the repeated task addition message into a new method, newTaskAddedMsg(), to reduce duplication when adding different types of tasks (Task, Deadline, Event).

This change ensures better code reusability and improves adherence to coding standards by eliminating repetitive code. The new method is now used in all addTask() methods to output a consistent message for each task type.

Commit:
Add assert statements to ensure user command input is never null
Added assert statements across multiple classes to verify that the user command input is not null. This ensures that the program does not proceed with invalid or unexpected inputs, preventing potential null pointer exceptions.

These assertions act as safety checks during development and testing, adhering to defensive programming practices and improving code robustness. In a production environment, these would be optimized or replaced with proper exception handling.

Commit:
Add assert line to Storage.java

sheenkerr committed:
Merge pull request #2 from sheenkerr/branch-A-CodeQuality

Branch a code quality

sheenkerr committed:
Merge pull request #3 from sheenkerr/branch-A-Assertions

Branch a assertions

Commit:
Add comparator to all task classes for sorting logic

Commit:
Add sort command into UI class

Commit:
Add sort function into TaskList.java

Commit:
Merge branch "extension"

Commit:
Apply Checkstyle formatting to TaskList and Storage classes

Updated the TaskList and Storage classes to conform to the project's Checkstyle formatting guidelines. This includes adjustments to indentation, spacing, and line breaks to ensure consistent code style and readability across the codebase.

By following Checkstyle rules, the code is now more maintainable and aligned with coding standards, which facilitates easier collaboration and future development.

Commit:
Add title and icon to java app

Commit:
Fix errors in parsing todo command

Commit:
Add ui.png to docs

Commit:
Update mark task to mark and unmark tasks

Commit:
Update TaskList to adhere to checkstyle

Commit:
Prompt user correct format when adding tasks

Commit:
Update findTask to check for input errors

Commit:
Update references for sorting logic

Commit:
Add custom exception error handling

Commit:
Update the User Guide

Commit:
Update the User Guide

Commit:
Update the GUI to close app upon exit

Commit:
Update variable name in Task to follow convention

Commit:
Update javadoc to follow convention

Commit:
Update classes to follow checkstyle convention

Commit:
Add reference to images used

Commit:
Update names in unit test to fit convension

Commit:
Add function to TaskList for testing purposes

Commit:
Add JUnit test for TaskList

Commit:
Add JUnit test for Parser

Commit:
Fix error preventing jar file from being executed

Commit:
Fix url formating in README
